interceptors/logging/logging.go

```diff
@@ -37,46 +35,110 @@ var (
-type ErrorToCode func(err error) codes.Code
+type FieldsIter interface {
+	Next() bool
+	At() (k, v string)
+}
+
+type iter struct {
+	f Fields
+	i int
+}
+
+func (i iter) Next() bool {
+	if i.i >= len(i.f) {
+		return false
+	}
+
+	i.i += 2
+	return i.i < len(i.f)
+}
+
+func (i iter) At() (k, v string) {
+	if i.i < 0 || i.i >= len(i.f) {
+		return "", ""
+	}
 
-func DefaultErrorToCode(err error) codes.Code {
-	return status.Code(err)
+	if i.i+1 == len(i.f) {
+		// Non even number of elements, add empty string.
+		return i.f[i.i], ""
+	}
+	return i.f[i.i], i.f[i.i+1]
 }
 
-// Decider function defines rules for suppressing any interceptor logs
-type Decider func(fullMethodName string, err error) Decision
+// AppendUnique returns fields which is the union of all keys with the added values having lower priority.
+func (f Fields) AppendUnique(add Fields) Fields {
+	if len(add) == 0 {
+		return f
+	}
+
+	existing := map[string]struct{}{}
+	i := f.Iter()
+	for i.Next() {
+		k, _ := i.At()
+		existing[k] = struct{}{}
+	}
+
+	n := make(Fields, len(f), len(f)+len(add))
+	copy(n, f)
 
-// DefaultDeciderMethod is the default implementation of decider to see if you should log the call
-// by default this if always true so all calls are logged
-func DefaultDeciderMethod(_ string, _ error) Decision {
-	return LogStartAndFinishCall
+	a := add.Iter()
+	for a.Next() {
+		k, v := a.At()
+		if _, ok := existing[k]; ok {
+			continue
+		}
+		n = append(n, k, v)
+	}
+	return n
 }
 
-// ServerPayloadLoggingDecider is a user-provided function for deciding whether to log the server-side
-// request/response payloads
-type ServerPayloadLoggingDecider func(ctx context.Context, fullMethodName string, servingObject interface{}) bool
+// ExtractFields returns logging.Fields object from the Context.
+// Logging interceptor adds fields into context when used.
+// If no one injected fields before ExtractFields returns empty Fields.
+//
+// It's useful for server implementations to use this method to instantiate request logger for consistent fields (e.g request-id/tracing-id).
+func ExtractFields(ctx context.Context) Fields {
+	t, ok := ctx.Value(fieldsCtxMarkerKey).(Fields)
+	if !ok {
+		return Fields{}
+	}
+	n := make(Fields, len(t))
+	copy(n, t)
+	return n
+}
 
-// ClientPayloadLoggingDecider is a user-provided function for deciding whether to log the client-side
-// request/response payloads
-type ClientPayloadLoggingDecider func(ctx context.Context, fullMethodName string) bool
+// InjectFields allows to add logging.Fields that will be used logging interceptor in the path of given context (if any).
+func InjectFields(ctx context.Context, f Fields) context.Context {
+	return context.WithValue(ctx, fieldsCtxMarkerKey, ExtractFields(ctx).AppendUnique(f))
+}
 
-// JsonPbMarshaller is a marshaller that serializes protobuf messages.
-type JsonPbMarshaler interface {
+// JsonPBMarshaller is a marshaller that serializes protobuf messages.
+type JsonPBMarshaller interface {
```

Comment:
The US English spelling of this is `Marshaler`. This naming is consistent with the gRPC gateway (https://pkg.go.dev/github.com/grpc-ecosystem/grpc-gateway/v2/runtime#JSONPb), but note that Go protobuf APIv2 calls this `protojson`.
```suggestion
// JsonPBMarshaler is a marshaler that serializes protobuf messages.
type JsonPBMarshaler interface {
```